types/src/v0/impls/state.rs

```diff
@@ -267,22 +267,54 @@ fn charge_fee(
 
 /// Validate builder account by verifying signature
 fn validate_builder_fee(proposed_header: &Header) -> Result<(), BuilderValidationError> {
-    // Beware of Malice!
+    // verify signatures
+
+    // TODO remove.
+    // Before this is finalized we will need a signature per fee_info
     let signature = proposed_header
         .builder_signature()
         .ok_or(BuilderValidationError::SignatureNotFound)?;
-    let fee_amount = proposed_header.fee_info().amount().as_u64().ok_or(
-        BuilderValidationError::FeeAmountOutOfRange(proposed_header.fee_info().amount()),
-    )?;
-
-    // verify signature
-    if !proposed_header.fee_info().account.validate_fee_signature(
-        &signature,
-        fee_amount,
-        proposed_header.metadata(),
-        &proposed_header.payload_commitment(),
-    ) {
-        return Err(BuilderValidationError::InvalidBuilderSignature);
+
+    match proposed_header {
+        Header::V3(proposed_header) => {
```

Comment:
Is there a better way to do this. If we were to match by value how would we inform the compiler which version of header we have?

Reply:
Per discussion in Zulip it appears these we are enforcing data structure upgrades on de-serialization, so we don't need the branches for different versions here. We can just handle the latest version.